File: py-polars/tests/benchmark/groupby-datagen.R
# Rscript groupby-datagen.R 1e7 1e2 0 0 ## 1e7 rows, 1e2 K, 0% NAs, random order
# Rscript groupby-datagen.R 1e8 1e1 5 1 ## 1e8 rows, 10 K, 5% NAs, sorted order
args = commandArgs(TRUE)

pretty_sci = function(x) {
  tmp<-strsplit(as.character(x), "+", fixed=TRUE)[[1L]]
  if(length(tmp)==1L) {
    paste0(substr(tmp, 1L, 1L), "e", nchar(tmp)-1L)
  } else if(length(tmp)==2L){
    paste0(tmp[1L], as.character(as.integer(tmp[2L])))
  }
}

library(data.table)
N=as.integer(args[1L]); K=as.integer(args[2L]); nas=as.integer(args[3L]); sort=as.integer(args[4L])
stopifnot(nas<=100L, nas>=0L, sort%in%c(0L,1L))
set.seed(108)
cat(sprintf("Producing data of %s rows, %s K groups factors, %s NAs ratio, %s sort flag\n", pretty_sci(N), pretty_sci(K), nas, sort))
DT = list()
DT[["id1"]] = sample(sprintf("id%03d",1:K), N, TRUE)      # large groups (char)
DT[["id2"]] = sample(sprintf("id%03d",1:K), N, TRUE)      # small groups (char)
DT[["id3"]] = sample(sprintf("id%010d",1:(N/K)), N, TRUE) # large groups (char)
DT[["id4"]] = sample(K, N, TRUE)                          # large groups (int)
DT[["id5"]] = sample(K, N, TRUE)                          # small groups (int)
DT[["id6"]] = sample(N/K, N, TRUE)                        # small groups (int)
DT[["v1"]] =  sample(5, N, TRUE)                          # int in range [1,5]
DT[["v2"]] =  sample(15, N, TRUE)                         # int in range [1,15]
DT[["v3"]] =  round(runif(N,max=100),6)                   # numeric e.g. 23.574912
setDT(DT)
if (nas>0L) {
  cat("Inputting NAs\n")
  for (col in paste0("id",1:6)) {
    ucol = unique(DT[[col]])
    nna = as.integer(length(ucol) * (nas/100))
    if (nna)
      set(DT, DT[.(sample(ucol, nna)), on=col, which=TRUE], col, NA)
    rm(ucol)
  }
  nna = as.integer(nrow(DT) * (nas/100))
  if (nna) {
    for (col in paste0("v",1:3))
      set(DT, sample(nrow(DT), nna), col, NA)
  }
}
if (sort==1L) {
  cat("Sorting data\n")
  setkeyv(DT, paste0("id", 1:6))
}
file = sprintf("G1_%s_%s_%s_%s.csv", pretty_sci(N), pretty_sci(K), nas, sort)
cat(sprintf("Writing data to %s\n", file))
fwrite(DT, file)
cat(sprintf("Data written to %s, quitting\n", file))
if (!interactive()) quit("no", status=0)

File: py-polars/tests/benchmark/data/groupby_datagen.py
"""
Script to generate data for benchmarking group-by operations.

Data generation logic was adapted from the H2O.ai benchmark.
The original R script is located here:
https://github.com/h2oai/db-benchmark/blob/master/_data/groupby-datagen.R

Examples
--------
10 million rows, 100 groups, no nulls, random order:
$ python groupby_datagen.py 1e7 1e2 --null-percentage 0

100 million rows, 10 groups, 5% nulls, sorted:
$ python groupby_datagen.py 1e8 1e1 --null-percentage 5 --sorted
"""

import argparse

import numpy as np
from numpy.random import default_rng

import polars as pl

SEED = 0
rng = default_rng(seed=SEED)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Generate data for benchmarking group-by operations"
    )

    parser.add_argument("rows", type=float, help="Number of rows")
    parser.add_argument("groups", type=float, help="Number of groups")
    parser.add_argument(
        "-n",
        "--null-percentage",
        type=int,
        default=0,
        choices=range(1, 101),
        metavar="[0-100]",
        help="Percentage of null values",
    )
    parser.add_argument(
        "-s",
        "--sort",
        action="store_true",
        help="Sort the data by group",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        default=None,
        help="Output filename",
    )

    args = parser.parse_args()

    # Convert arguments to appropriate types
    N = int(args.rows)
    K = int(args.groups)
    null_ratio = args.null_percentage / 100

    print(
        f"Producing data of {N} rows, {K} K groups factors, {args.null_percentage}% nulls, sorted: {args.sort}"
    )

    df = pl.DataFrame(
        {
            "id1": rng.choice([f"id{str(i).zfill(3)}" for i in range(1, K + 1)], N),
            "id2": rng.choice([f"id{str(i).zfill(3)}" for i in range(1, K + 1)], N),
            "id3": rng.choice(
                [f"id{str(i).zfill(10)}" for i in range(1, int(N / K) + 1)], N
            ),
            "id4": rng.choice(range(1, K + 1), N),
            "id5": rng.choice(range(1, K + 1), N),
            "id6": rng.choice(range(1, int(N / K) + 1), N),
            "v1": rng.choice(range(1, 6), N),
            "v2": rng.choice(range(1, 16), N),
            "v3": np.round(rng.uniform(0, 100, N), 6),
        }
    )

    if null_ratio > 0.0:
        print("Setting nulls")
        df = set_nulls(df, null_ratio)

    if args.sort:
        print("Sorting data")
        df = df.sort(c for c in df.columns if c.startswith("id"))

    write_data(df, args)


def set_nulls(df: pl.DataFrame, null_ratio: float) -> pl.DataFrame:
    """Set null values according to the given ratio."""

    def set_nulls_var(s: pl.Series, ratio: float) -> pl.Series:
        """Set Series values to null according to the given ratio."""
        len = s.len()
        n_null = int(ratio * len)
        if n_null == 0:
            return s

        indices = rng.choice(len, size=n_null, replace=False)
        return s.scatter(indices, None)

    def set_nulls_group(s: pl.Series, ratio: float) -> pl.Series:
        """Set Series unique values to null according to the given ratio."""
        uniques = s.unique()
        n_null = int(ratio * uniques.len())
        if n_null == 0:
            return s

        to_replace = rng.choice(uniques, size=n_null, replace=False)
        return (
            s.to_frame()
            .select(
                pl.when(pl.col(s.name).is_in(to_replace))
                .then(None)
                .otherwise(pl.col(s.name))
                .alias(s.name)
            )
            .to_series()
        )

    return df.with_columns(
        set_nulls_group(s, null_ratio)
        if s.name.startswith("id")
        else set_nulls_var(s, null_ratio)
        for s in df.get_columns()
    )


def write_data(df: pl.DataFrame, args: argparse.Namespace) -> None:
    def format_int(i: int) -> str:
        base, exp = f"{i:e}".split("e")
        base = f"{float(base):g}"
        exp = int(exp)
        return f"{base}e{exp}"

    if args.output is not None:
        filename = args.output
    else:
        filename = f"G1_{format_int(args.rows)}_{format_int(args.groups)}_{args.null_percentage}_{int(args.sort)}.csv"

    print(f"Writing data to {filename}")
    df.write_csv(filename)


if __name__ == "__main__":
    main()
